staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -138,6 +150,78 @@ func (w *watcher) createWatchChan(ctx context.Context, key string, rev int64, re
 	return wc
 }
 
+// isInitialEventsEndBookmarkRequired since there is no way to directly set
+// opts.ProgressNotify from the API and the etcd3 impl doesn't support
+// notification for external clients we simply return initialEventsEndBookmarkRequired
+// to only send the bookmark event after the initial list call.
+//
+// see: https://github.com/kubernetes/kubernetes/issues/120348
+func (w *watcher) isInitialEventsEndBookmarkRequired(opts storage.ListOptions) (bool, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.WatchList) {
+		return false, nil
+	}
+	if opts.SendInitialEvents == nil || !*opts.SendInitialEvents || !opts.Predicate.AllowWatchBookmarks {
+		return false, nil
+	}
+	if err := w.validateRequiredFieldsForWatchListFeature(); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// getStartWatchResourceVersion returns a ResourceVersion
+// the watch will be started from.
+// Depending on the input parameters the semantics of the returned ResourceVersion are:
+//   - start at Exact (return resourceVersion)
+//   - start at Most Recent (return an RV from etcd)
+func (w *watcher) getStartWatchResourceVersion(ctx context.Context, resourceVersion int64, opts storage.ListOptions) (int64, error) {
+	if resourceVersion > 0 {
+		return resourceVersion, nil
+	}
+	if !utilfeature.DefaultFeatureGate.Enabled(features.WatchList) {
+		return resourceVersion, nil
+	}
+	if opts.SendInitialEvents == nil || *opts.SendInitialEvents {
+		return resourceVersion, nil
+	}
+	if err := w.validateRequiredFieldsForWatchListFeature(); err != nil {
```

Comment:
I think we're going to far in defensive programming here.

In validateRequiredFields... you're not validating the user input - you're checking for programmers errors.
I would just remove it - this is complicating the code for not value.

We should use validation to validate user inputs...

Reply:
I don't like it but okay, I can remove it.